publications/radstack/radstack.tex

```diff
@@ -0,0 +1,1148 @@
+to handle complex nested filter sets is what enables Druid to drill into data
+at any depth. 
+
+Druid supports many types of aggregations including sums on floating-point and
+integer types, minimums, maximums, and complex aggregations such as cardinality
+estimation and approximate quantile estimation. The results of aggregations can
+be combined in mathematical expressions to form other aggregations. It is
+beyond the scope of this paper to fully describe the query API but more
+information can be found
+online\footnote{\href{http://druid.io/docs/latest/Querying.html}{http://druid.io/docs/latest/Querying.html}}. 
+
+As of this writing, a join query for Druid is not yet implemented.  Although
+Druid’s storage format would allow for the implementation of joins (there is no
+loss of fidelity for columns included as dimensions), implementation time is
+costly and a strong enough use case has not yet arised. The reasons for this
+decision are generally two-fold. 
+
+1. Scaling join queries has been, in our professional experience, a constant
+bottleneck of working with distributed databases. 
+
+2. The incremental gains in functionality are perceived to be of less value
+than the anticipated problems with managing highly concurrent, join-heavy
+workloads. 
+
+A join query is essentially the merging of two or more streams of data based on
+a shared set of keys. The primary high-level strategies for join queries we are
+aware of are a hash-based strategy or a sorted-merge strategy. The hash-based
+strategy requires that all but one data set be available as something that
+looks like a hash table, a lookup operation is then performed on this hash
+table for every row in the “primary” stream. The sorted-merge strategy assumes
+that each stream is sorted by the join key and thus allows for the incremental
+joining of the streams. Each of these strategies, however, requires the
+materialization of some number of the streams either in sorted order or in a
+hash table form.
+
+When all sides of the join are significantly large tables (> 1 billion
+records), materializing the pre-join streams requires complex distributed
+memory management. The complexity of the memory management is only amplified by
+the fact that we are targeting highly concurrent, multi-tenant workloads. For
+these reasons, we’ve elected to do joins at the processing layer for the time
+being.
+
+\section{The Processing Layer}
+\label{sec:processing}
+Although Druid can ingest events that are streamed in one at a time, data must
+be denormalized beforehand as Druid cannot yet support
+join queries. Furthermore, real world data must often be transformed before it
+is usable by an application.
+
+\subsection{Stream Processing}
+\label{sec:streamprocessing}
+Stream processors provide infrastructure to develop processing logic for
+unbounded sequences of messages. We use Apache Samza as our stream processor,
+although other technologies are viable alternatives. Samza provides an API to
```

Comment:
"are viable alternatives (indeed, we initially chose Storm, but have since switched to Samza)"?